File: js/linklist.js
const readline = require('readline-sync');

class Node {

    constructor() {
        this.data;
        this.next = null;
    }
}

class linklist {

    constructor() {
        this.head = null;
        this.new_node = null;
        this.current = null;
        this.size = 0;
    }

    append(element) {
                
        this.new_node = new Node();
        this.new_node.data = element;

        if(this.head===null) {
            this.head = this.new_node;
        }
        else {
            this.current = this.head;
            while(this.current.next) {
                this.current = this.current.next;
            }
            this.current.next = this.new_node;
        }
        this.size++;
    }

    insert_begening(element) {

        this.new_node = new Node(); // created new node
        this.new_node.data = element; // put data into node
        this.new_node.next = this.head; // new_node -> next = head
        this.head = this.new_node; // head = new_node
        this.size++;
    }

    remove(pos) {

        var pos = Number(pos);
        var curr,prev;
        curr = this.head;
        
        if(pos>this.size) {
            console.log("invalid");
        }
        else if(pos===1) {
            console.log("removed element : "+curr.data);
            curr = curr.next;
            this.head = curr;
        }
        else {
            for(var i=0;i<pos-1;i++) {
                prev = curr
                curr = curr.next;
            }
            prev.next = curr.next;
            console.log("removed element : "+curr.data);
        }
        this.size--;

    }

    insertAt(pos) {
        
        var pos = Number(pos);
        var curr,prev;

        this.new_node = new Node();
        var element = readline.question("enter the data into node : ");
        this.new_node.data = element;

        if(pos==1) {
            // insert at begining
            this.new_node.next = this.head;
            this.head = this.new_node;
        }
        else {
            // at any other position
            curr = this.head;
            for(var i=0;i<pos-1;i++) {
                prev = curr;
                curr = curr.next;
            }
            this.new_node.next = curr;
            prev.next = this.new_node;
        }
        this.size++;
    }

    search_element() {
        
        var curr;

        var element = readline.question("enter the element to search : ");

        curr = this.head;
        var loc = 0,flag=0;
        while(curr) {
            if(curr.data===element) {
                flag=1;
            }
            loc++;
            curr = curr.next;
        }
        if(flag!=0) {
            console.log("element found at location = "+loc);
        }
        else {
            console.log("element not found!");
        }
        
    }

    display() {

        var temp = null;
        temp = this.head;
        while(temp.next) {
            console.log(temp.data);
            temp = temp.next;
        }
        console.log(temp.data);
    }


}

// creating object of class
var obj = new linklist();

do {

    console.log("1. add at end : ");
    console.log("2. insert at begining : ");
    console.log("3. remove from pos : ")
    console.log("4. insert at pos  ");
    console.log("5. display : ");
    console.log("6. search element : ");
    console.log("0. EXIT ");

    var choice = readline.question();

    switch(choice) {
        case '1' :
                var data = readline.question("enter data into node : ");
                obj.append(data);
                break;
        
        case '2' :
                var data = readline.question("enter the data into node : ");
                obj.insert_begening(data);
                break;

        case '3' :
                var pos = readline.question("enter the position : ");
                obj.remove(pos);
                break;

        case '4' :
                var pos = readline.question("enter the position : ");
                obj.insertAt(pos);
                break;

        case '5' : 
                obj.display();
                break;

        case '6' : 
                obj.search_element();
                break;
    }

}while(choice!='0');

File: mayur/linklist.cpp
#include<iostream>

using namespace std;

typedef struct Node {
    int data;
    Node *next;
} Node;

Node *newnode = NULL;
Node *head = NULL;
Node *curr = NULL;
Node *previ = NULL;
int count = 0;

// insert node[at end]
void append() {

    int element;

    cout<<"enter element : "<<endl;
    cin>>element;

    newnode = new Node();
    newnode->data = element;

    if(!newnode) {
        cout<<"linklist full."<<endl;
        exit(0);
    }

    else {

        if(head==NULL) {
            head = newnode;
        }

        else {
            curr = head;
            while(curr->next!=NULL) {
                curr = curr->next;
            }
            curr->next = newnode;
        }
        count++;
    }

    cout<<"inserted."<<endl;
}

// insert node[at begining]
void insertAtBeg() {

    int element;

    cout<<"enter element : "<<endl;
    cin>>element;

    newnode = new Node();
    newnode->data = element;

    if(!newnode) {
        cout<<"link list is full"<<endl;
        exit(0);
    }

    else {
        // link list is not full; we can insert the node
        if(head==NULL) {
            // link list is empty;
            head = newnode;
        }
        else {
            newnode->next = head;
            head = newnode;
        }
        count++;
    }
    cout<<"inserted."<<endl;
}

void insertAt(int pos) {

    int element;

    cout<<"enter element : "<<endl;
    cin>>element;

    // create newnode
    newnode = new Node(); // memory allocation for newnode pointer
    newnode->data = element;

    if(!newnode) {
        cout<<"linklist is full"<<endl;
        exit(0);
    }

    else {

        if(head==NULL) {
            cout<<"linklist is empty.. adding first node"<<endl;
            head = newnode;
        }

        else {
            curr = head;
            for(int i=0;i<pos-1;i++) {
                previ = curr;
                curr = curr->next;
            }
            previ->next = newnode;
            newnode->next = curr;
        }

        count++;
    }
    cout<<"inserted."<<endl;
}

// remove from end
int removeFromEnd() {

    int element;

    if(count==0) {
        cout<<"linklist is empty"<<endl;
        exit(0);
    }

    curr = head;
    while(curr->next!=NULL) {
        previ = curr;
        curr = curr->next;
    }
    element = curr->data;
    previ->next = NULL;
    free(curr);
    count--;
    return element;
}

// remove from begining
int removeFromBeg() {

    int element;

    if(count==0) {
        cout<<"linklist is empty"<<endl;
        exit(0);
    }

    curr = head;
    head = head->next;
    element = curr->data;
    curr->next = NULL;
    free(curr);
    count--;
    return element;
}

int removeAt(int pos) {

    int element;

    if(count==0) {
        cout<<"linklist is empty"<<endl;
        exit(0);
    }

    curr = head;
    for(int i=0;i<pos-1;i++) {
        previ = curr;
        curr = curr->next;
    }

    previ->next = curr->next;
    element = curr->data;
    curr->next = NULL;
    free(curr);
    count--;
    return element;
}

// search element
void search(int element) {
    
    int flag = 0;
    int loc = 0;
    if(element == head->data) {
        cout<<"found"<<endl;
    }

    else {

        curr = head;
        while(curr->next!=NULL) {
            if(curr->data == element) {
                flag=1;
                break;
            }
            loc++;
            curr = curr->next;

        }

        if(flag==1) {
            cout<<"element found at location : "<<loc+1<<endl;
        }
        else {
            cout<<"not found"<<endl;
        }
    }

}

// display link list 
void display() {

    curr = head;
    while(curr->next!=NULL) {
        cout<<curr->data<<endl;
        curr = curr->next;
    }
    cout<<curr->data<<endl;
}

// driver code
int main () {

    int choice;
    int pos;
    int element;

    do {

        cout<<"1. insert node[at end] : "<<endl;
        cout<<"2. insert node[at begining]"<<endl;
        cout<<"3. insert node[at pos]"<<endl;
        cout<<"4. remove node[at end] : "<<endl;
        cout<<"5. remove node[at begining]"<<endl;
        cout<<"6. remove node[at pos]"<<endl;
        cout<<"7. display node : "<<endl;
        cout<<"8. search node : "<<endl;
        cout<<"0. EXIT : "<<endl;
        
        cin>>choice;

        switch(choice) {

            case 1 :
            append();
            break;

            case 2 :
            insertAtBeg();
            break;

            case 3 :
            cout<<"enter pos : "<<endl;
            cin>>pos;
            insertAt(pos);
            break;

            case 4 : 
            element = removeFromEnd();
            cout<<element<<" removed."<<endl;
            break;              

            case 5 : 
            element = removeFromBeg();
            cout<<element<<" removed"<<endl;
            break;

            case 6 : 
            cout<<"enter pos : "<<endl;
            cin>>pos;
            element = removeAt(pos);
            cout<<element<<" removed."<<endl;   
            break;

            case 7 : 
            display();
            break; 
            
            case 8 :
            cout<<"enter element : "<<endl;
            cin>>element;
            search(element);
            break;
        }

    }while(choice!=0);

}
